### PR TITLE
DAOS-13194 gurt: environment APIs hook

### PR DESCRIPTION
**Pull request by bfaccini**
In order to prevent known race to occur due to lack of locking in Glibc environment APIs (getenv()/[uns]setenv()/ putenv()/clearenv()), they have been overloaded and strengthened in Gurt with hooks now all using a common lock/mutex.

Libgurt is the preferred place for this as it is the lowest layer in DAOS, so it will be the earliest to be loaded and will ensure the hook to be installed as early as possible and could prevent usage of LD_PRELOAD.

This will address the main lack of multi-thread protection in the Glibc APIs but do not handle all unsafe use-cases (like the change/removal of an env var when its value address has already been grabbed by a previous getenv(), ...).

Required-githooks: true

### Before requesting gatekeeper:

* [ ] Two review approvals and any prior change requests have been resolved.
* [ ] Testing is complete and all tests passed or there is a reason documented in the PR why it should be force landed and forced-landing tag is set.
* [ ] `Features:` (or `Test-tag*`) commit pragma was used or there is a reason documented that there are no appropriate tags for this PR.
* [ ] Commit messages follows the guidelines outlined [here](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments).
* [ ] Any tests skipped by the ticket being addressed have been run and passed in the PR.

### Gatekeeper:

* [ ] You are the appropriate gatekeeper to be landing the patch.
* [ ] The PR has 2 reviews by people familiar with the code, including appropriate watchers.
* [ ] Githooks were used. If not, request that user install them and check copyright dates.
* [ ] Checkpatch issues are resolved.  Pay particular attention to ones that will show up on future PRs.
* [ ] All builds have passed.  Check non-required builds for any new compiler warnings.
* [ ] Sufficient testing is done. Check feature pragmas and test tags and that tests skipped for the ticket are run and now pass with the changes.
* [ ] If applicable, the PR has addressed any potential version compatibility issues.
* [ ] Check the target branch.   If it is master branch, should the PR go to a feature branch?  If it is a release branch, does it have merge approval in the JIRA ticket.
* [ ] Extra checks if forced landing is requested
  * [ ] Review comments are sufficiently resolved, particularly by prior reviewers that requested changes.
  * [ ] No new NLT or valgrind warnings.  Check the classic view.
  * [ ] Quick-build or Quick-functional is not used.
* [ ] Fix the commit message upon landing. Check the standard [here](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments). Edit it to create a single commit. If necessary, ask submitter for a new summary.

### CODE REVIEW

File: src/gurt/misc.c
Comment (frostedcmos) [inline src/gurt/misc.c:1241]: Seems like this block is repeated across the code 5 times, once for each call, with the only difference being the symbol being looked up and used. Can this be simplified to use a shared function across those calls instead?
Something like

```
char *getenv(const char *name)
{
   return shared_call(&real_getenv, "getenv");
}

char *setenv(const char *name)
{
   return shared_call(&real_setenv, "setenv");
}
```

etc...

File: src/gurt/misc.c
Comment (jolivier23) [inline src/gurt/misc.c:1228]: dlsym returns a void *. You don't need to cast it

File: src/gurt/misc.c
Comment (jolivier23) [inline src/gurt/misc.c:1250]: I might suggest using pthread_once and initializing all of the hooks at once?

File: src/gurt/misc.c
Comment (bfaccini) [reply src/gurt/misc.c:1222]: As I have already indicated in DAOS-13194 JiRA/ticket :
> I have manually verified that this approach works for the daos/dfuse/dmg/daos_agent/daos_server/daos_engine/daos_server_helper binaries.

File: src/gurt/misc.c
Comment (bfaccini) [reply src/gurt/misc.c:1228]: Ah... and do you know where does this rule come from ? Early Clang definitions, Kernel developers, ...??

File: src/gurt/misc.c
Comment (bfaccini) [reply src/gurt/misc.c:1240]: > If you separate init from this, the lock here should really only protect the real call.

Well, initialisation of the "real_..." variables is racy too, but may be I can switch to using "atomic" variables then ?!

File: src/gurt/misc.c
Comment (bfaccini) [reply src/gurt/misc.c:1250]: Ah yes, you are right !!

File: src/gurt/misc.c
Comment (bfaccini) [reply src/gurt/misc.c:1260]: Finally, I have switched to using Atomics for the initialisation part and also switched to rw-lock usage (instead of simple mutex, mainly to allow concurrent getenv()s) for the exception part.

**Comment by bfaccini** [comment]
The set of errors that have occurred during "Functional Hardware Medium Verbs Provider" stage again appear to be related to DAOS-12883/DAOS-13488 ... Will merge with recent master to see if it can fix.

**Comment by bfaccini** [comment]
> Even after merging with last master, the 3 errors that have occurred during "Functional Hardware Medium Verbs Provider" stage again appear to be related to DAOS-12883/DAOS-13488 ...

BTW, simply re-running the "Functional Hardware Medium Verbs Provider" stage has worked finally...

**Review by frostedcmos — APPROVED** [review]

File: src/gurt/misc.c
Comment (soumagne) [inline src/gurt/misc.c:1261]: I feel like you want to use contructors instead using `__attribute__((constructor))` ? I'm not sure why you'd use `pthread_once` etc.

File: src/gurt/misc.c
Comment (bfaccini) [reply src/gurt/misc.c:1243]: Well, for some (obscure ?) reason "libc.so" is no longer a symbolic link, but a text file (apparently a "dynamic loader script"...) causing dlopen() to fail with "invalid ELF header".

**Comment by bfaccini** [comment]
> I'm not particularly enthusiastic on having this kind of redefinition,

This is still open, first attempt to preallocate bunch of environment array slots ([PR-12166](https://github.com/daos-stack/daos/pull/12166)) has been abandoned because not working finally, this is a second try to fix DAOS-13194.

> I feel like if there is any non-threadsafe use with setenv(), it's potentially masking another problem that is calling setenv multiple times ? which seems bad ?

I am sorry, but I don't get where you think current approach could fail ?

**Review by jolivier23 — DISMISSED** [review]

**Comment by jolivier23** [comment]
> I'm not particularly enthusiastic on having this kind of redefinition, I feel like if there is any non-threadsafe use with setenv(), it's potentially masking another problem that is calling setenv multiple times ? which seems bad ?

I'm trying to understand why that's an issue.   The real fix for all of these issues is to stop using environment variables to control behavior.

File: src/gurt/misc.c
Comment (bfaccini) [reply src/gurt/misc.c:1238]: > so this doesn't find libc version? Seems like it should always find it since libc should be linked.

I was also convinced that libc should always be binded, but if you look to the CI failures of the 2nd commit in this PR (https://build.hpdd.intel.com/blue/organizations/jenkins/daos-stack%2Fdaos/detail/PR-12220/2/pipeline/777/) there are multiple ftests that could fail because dlsym() was not able to resolve "getenv" symbol. May be this is a python only issue...

**Review by soumagne — DISMISSED** [review]

File: src/gurt/misc.c
Comment (daltonbohning) [reply src/gurt/misc.c:1235]: @bfaccini Looks like this needs to be resolved

**Review by jolivier23 — COMMENTED** [review]
can you address the two style warnings?

File: src/gurt/misc.c
Comment (bfaccini) [reply src/gurt/misc.c:1235]: done

**Comment by bfaccini** [comment]
> can you address the two style warnings?

Two warnings ? I only see one for the trailing space at line 1235. Can you point me the 2nd ?

**Review by soumagne — DISMISSED** [review]

**Review by jolivier23 — APPROVED** [review]

**Comment by ashleypittman** [comment]
This has a conflict.

**Comment by bfaccini** [comment]
> This has a conflict.

Ok, just merged with latest master ... really sorry for my reviewers :-(

**Review by ashleypittman — APPROVED** [review]

**Review by frostedcmos — APPROVED** [review]

**Comment by bfaccini** [comment]
After a merge with latest master, a forced test rerun, a new merge with latest master, I have been able to get 2 consecutive and successful CI sessions with no more apparently unrelated unit-tests errors (due to "nvme_control_ctests" unexpectedly missing) !!...

File: src/gurt/misc.c
Comment (ashleypittman) [reply src/gurt/misc.c]: `void *` will alias to anything, it never needs a cast.

**Pull request merged**
